realm/realm-library/src/main/java/io/realm/internal/objectstore/OsObjectBuilder.java

```diff
@@ -0,0 +1,415 @@
+/*
+ * Copyright 2018 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm.internal.objectstore;
+
+import java.util.Date;
+import java.util.List;
+import java.util.Set;
+
+import io.realm.ImportFlag;
+import io.realm.MutableRealmInteger;
+import io.realm.RealmList;
+import io.realm.RealmModel;
+import io.realm.internal.NativeContext;
+import io.realm.internal.OsSharedRealm;
+import io.realm.internal.RealmObjectProxy;
+import io.realm.internal.Table;
+import io.realm.internal.UncheckedRow;
+
+/**
+ * This class is a wrapper around building up object data for calling `Object::create()`
+ * <p>
+ * Fill the object data by calling the various `addX()` methods, then create a new Object or update
+ * an existing one by calling {@link #createNewObject()} or {@link #updateExistingObject()}.
+ * <p>
+ * This class assumes it is only being used from within a write transaction. Using it outside one
+ * will result in undefined behaviour.
+ * <p>
+ * The native
+ * resources are created in the constructor of this class and destroyed when calling either of the
+ * above two methods.
+ * <p>
+ * <H1>Design thoughts</H1>
+ * <p>
+ * Ideally we would have sent all properties across in one JNI call, but the only way to do that would
+ * have been using two `Object[]` arrays which would have resulted in a ton of JNI calls back
+ * again for resolving the primitive values of boxed types (since JNI do not know about boxed
+ * primitives).
+ * <p>
+ * The upside of making a JNI call for each property is that we do minimal allocations on the Java
+ * side. Also each method call is fairly lightweight as no checks are performed compared to using
+ * Proxy setters and {@link io.realm.internal.UncheckedRow}. The only downside is the current need for
+ * sending the key as well. Hopefully we can change that to schema indices at some point.
+ * <p>
+ * There is quite a few variants we can attempt to optimize this, but at this point we lack data
+ * that can guide any architectural design and the only way to really find out is to build out each
+ * solution and benchmark it.
+ */
+public class OsObjectBuilder {
```

Comment:
I actually did that originally, but that put reclaiming resources to the mercy of the GC. This can potentially be problematic if you are copying in a lot of objects so I opted for the pattern where you manually reclaim the native resources. This is possible here since we always know precisely when we are done with the class.